atecc608a_se.c

```diff
@@ -240,17 +247,22 @@ static psa_status_t atecc608a_generate_key(psa_key_slot_number_t key_slot,
         return PSA_ERROR_NOT_SUPPORTED;
     }
 
-    if (p_pubkey_out != NULL && pubkey_out_size < ATCA_PUB_KEY_SIZE)
+    if (p_pubkey_out != NULL && pubkey_out_size < key_data_len)
     {
        return PSA_ERROR_BUFFER_TOO_SMALL;
     }
 
     ASSERT_SUCCESS_PSA(atecc608a_init());
-    ASSERT_SUCCESS(atcab_genkey(key_id, p_pubkey_out));
+
+    /* atcab_genkey returns concatenated x and y values, and the desired
+     * format is 0x04 + x + y. Start at &p_pubkey_out[1] and add a 0x04
+     * at p_pubkey_out[0]. */
```

Comment:
Would you agree that its worth making a static helper functions to do format translation? It looks like any function dealing with pubkeys will need shared format translation code. I'd imagine we need one for driver_to_user and user_to_driver.

Reply:
We can either leave it as is, calling the cryptoauthlib with `&p_pubkey_out[1]` and adding a `0x04`, or increase the complexity by calling it at `&p_pubkey_out[0]`, then calling an utility function that would `memmove` it and add `0x04` at the beginning.
Since all the translation does is essentially adding / stripping off `0x04`, I would rather leave it as it is right now.

Reply:
We shouldn't need to use `memmove()`.

Maybe something like this would help us avoid having to repeat ourselves.

```C
/* The driver works with pubkeys as concatenated x and y values, and the PSA
 * format for pubkeys is 0x04 + x + y. Always use a pubkey buffer in PSA
 * format, with enough space for the PSA format. To translate this buffer for
 * use with cryptoauthlib, use pubkey_for_driver(). To ensure the buffer is in
 * valid PSA format after cryptoauthlib operations, call pubkey_for_psa(). */
static uint8_t *pubkey_for_driver(uint8_t *data)
{
    return &data[1];
}

static void pubkey_for_psa(uint8_t *data)
{
    data[0] = 0x4;
}

/* ... */
{
    ASSERT_SUCCESS(atcab_genkey(key_id, pubkey_for_driver(p_pubkey_out)));
    pubkey_for_psa(p_pubkey_out);
}
/* ... */
```

Reply:
I think the code is less readable with these trivial helpers, since you have to read more code to figure out whether the buffer is processed correctly.

Reply:
You have to read less code in total, when reviewing the entire driver, because only one function contains this logic. You are less likely to make mistakes when handling the format translation, since you have known-good functions to use instead of copy-paste; and not using the helpers is obvious in code review.